Feature/sprkstep

balos1 opened the pull request:
Todo:
- [x] Merge https://github.com/sundials-codes/answers/pull/16 first
- [x] Update example output in repo with output from Jenkins

Rendered docs can be viewed here: https://sundials.readthedocs.io/en/feature-sprkstep/index.html

drreynolds commented on src/arkode/arkode_sprkstep_io.c:
```diff
@@ -0,0 +1,491 @@
+  ---------------------------------------------------------------*/
+int SPRKStepSetUserData(void* arkode_mem, void* user_data)
+{
+  ARKodeMem ark_mem          = NULL;
+  ARKodeSPRKStepMem step_mem = NULL;
+  int retval                 = 0;
+
+  /* access ARKodeSPRKStepMem structure */
+  retval = sprkStep_AccessStepMem(arkode_mem, "SPRKStepSetUserData", &ark_mem,
+                                  &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* set user_data in ARKODE mem */
+  retval = arkSetUserData(arkode_mem, user_data);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  return (ARK_SUCCESS);
```
I see nothing stepper-specific about this routine, so it could be simplified and moved to the preceding section.
```suggestion
  return (arkSetUserData(arkode_mem, user_data));
```

drreynolds commented on src/arkode/arkode_sprkstep_io.c:
```diff
@@ -0,0 +1,491 @@
+    step_mem->method = NULL;
+  }
+
+  step_mem->method = ARKodeSPRKStorage_LoadByName(method);
```
Is this routine guaranteed to succeed?  If not, then shouldn't we check `if (step_mem->method == NULL)` before returning `ARK_SUCCESS`?

drreynolds commented on src/arkode/arkode_sprkstep_io.c:
```diff
@@ -0,0 +1,491 @@
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  /* set user-provided value, or default, depending on argument */
+  if (ord <= 0) { step_mem->q = 4; }
+  else { step_mem->q = ord; }
```
Should this also check for an upper bound, and/or whether `ord` matches "allowable" positive values?

drreynolds commented on src/arkode/arkode_sprkstep_io.c:
```diff
@@ -0,0 +1,491 @@
+    return (flag);
+  }
+
+  /* print integrator parameters to file */
```
Did you consider also writing out the SPRK method coefficients?

balos1 replied:
We can, but we do not do this with the other steppers in the WriteParameters functions. Thoughts?

drreynolds replied:
It's really up to you for SPRKStep -- I think of these WriteParameters routines as purely being useful when debugging nasty issues.  If you add it here, then we can also consider adding those to ARKStep, ERKStep and MRIStep.

balos1 replied:
Ill leave it out here as we can write them via the Butcher table functions.

drreynolds reviewed (commented):
I've reviewed everything *except* `src/arkode/arkode_sprkstep.c`.  I'll look at that tomorrow.

drreynolds commented on src/arkode/arkode_sprkstep.c:
```diff
@@ -0,0 +1,781 @@
+#include <sundials/sundials_math.h>
+#include <sundials/sundials_nvector.h>
+#include <sundials/sundials_types.h>
+#include <sunnonlinsol/sunnonlinsol_newton.h>
```
```suggestion
```

drreynolds replied:
It looks like this file still includes `<sunnonlinsol/sunnoninsol_newton.h>`, but I don't see why that is ever needed here.

drreynolds commented on src/arkode/arkode_sprkstep.c:
```diff
@@ -0,0 +1,781 @@
+    return (NULL);
+  }
+
+  /* We use Lagrange interpolation by default otherwise extra RHS calls are
+     needed. This is because we cannot reuse the f2 RHS in TakeStep since it is
+     a staggered time step. Additionally, it seems Lagrange interpolation does
+     a better job of conservation. */
```
I'm not sure that we need so much of a justification for this default here.  Perhaps something simpler:
```suggestion
  /* SPRKStep uses Lagrange interpolation by default, since Hermite is less compatible with these methods. */
```

drreynolds commented on src/arkode/arkode_sprkstep.c:
```diff
@@ -0,0 +1,781 @@
+int SPRKStepEvolve(void* arkode_mem, realtype tout, N_Vector yout,
+                   realtype* tret, int itask)
+{
+  /* unpack ark_mem, call arkGetDky, and return */
```
```suggestion
  /* unpack ark_mem, call arkEvolve, and return */
```

drreynolds commented on src/arkode/arkode_sprkstep.c:
```diff
@@ -0,0 +1,781 @@
+  case ARK_FULLRHS_END:
+  case ARK_FULLRHS_OTHER:
+
+    retval = sprkStep_f1(step_mem, t, y, f, ark_mem->user_data);
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, "ARKODE::SPRKStep",
+                      "SPRKStep_FullRHS", MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
+
+    retval = sprkStep_f2(step_mem, t, y, f, ark_mem->user_data);
+    if (retval != 0)
+    {
+      arkProcessError(ark_mem, ARK_RHSFUNC_FAIL, "ARKODE::SPRKStep",
+                      "SPRKStep_FullRHS", MSG_ARK_RHSFUNC_FAILED, t);
+      return (ARK_RHSFUNC_FAIL);
+    }
```
This only works if the problem is appropriately partitioned.  I don't recall the documentation giving any stern warnings against users attempting to provide non-partitioned RHS functions.

balos1 replied:
I added warnings.

drreynolds reviewed (changes requested):
Overall this looks great.  I tried to give "suggestions" for all easily-fixable items to streamline the update process.

Steven-Roberts commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1462 @@
+   be solved using the SPRKStep time-stepping module in ARKODE.
+
+   :param f1: the name of the C function (of type :c:func:`ARKRhsFn()`) defining :math:`f_1(q,t) = \frac{\partial V(q,t)}{\partial q}`
+   :param f2:  the name of the C function (of type :c:func:`ARKRhsFn()`) defining :math:`f_2(p) = \frac{\partial T(p)}{\partial p}`
```
Similarly, `f1` and `f2` should fill in disjoint sets of variables in `ydot` if I understand correctly. And `ydot` has length equal to the full dimension of the ODE? This documentation implies that `f1` and `f2` return subvectors for the `p` and `q` components only.

balos1 replied:
I updated the documentation to clarify that the vector returned is the full vector including `p` and `q`. If we wanted to make it so that it only the `p` and `q` components were sent to the Rhs functions, we would need to require the use of the ManyVector. 

There are no internal tests to ensure that f2 is independent of `t`.  What we send in is the `t` value sent into `f1` which is the `tn + ci * h`.

gardner48 replied:
`f1` is passed `ark_mem->tn + chati*ark_mem->h` and `f2` is passed `ark_mem->tn + ci*ark_mem->h`. Is it usually the case that `chati = ci`?

The notation for comments and documentation is inconsistent about having `f2` as `f(t, p)` or `f(p)`. I think I comment on most of these to add the `t` but maybe it better to leave it off. Thoughts?

balos1 replied:
I think we should add the `t` as it *should* work. `chati /= ci` in general. I see in the math section I flipped (typo) the use of \hat and no hat between the regular and incremental form.

balos1 replied:
Ill note that the original comment here was from before I made it possible to have a time-dependent f1 and f2.

drreynolds commented on doc/arkode/guide/source/Butcher.rst:
It looks like you removed the method-specific constructor functions from `arkode_sprk.h`, so we should remove those from this file as well (each block that states ", or by calling X.")

Steven-Roberts commented on src/arkode/arkode_sprk.c:
I might be able to help with deriving extra digits of precision if needed

balos1 replied:
I added the closed form for the 3rd order McLachlan. The 4th, 5th, 8th order McLachlan and 6th order Yoshida are all that are left with only double precision refined coefficients. @Steven-Roberts if you can help derive the extra digits for those, that would be great, although it seems it would take quite a bit of effort to get them based on https://www.tandfonline.com/doi/full/10.1080/10556780500140664.

Steven-Roberts replied:
Yoshida's coefficients to 40 digits:
```
{0.7845136104775572638194976338663498757768,0.2355732133593581336847931829785346016865,-1.177679984178871006946415680964315734639,1.315186320683911218884249728238862514352,-1.177679984178871006946415680964315734639,0.2355732133593581336847931829785346016865,0.7845136104775572638194976338663498757768}
```
I'll see about McLachlan methods tomorrow. Order conditions are pretty simple so I'm optimistic.

```
G = {g1, g2, g3, g4, g3, g2, g1};
sol = Solve[{
	Total[G] - 1,
	Total[G^3],
	Total[G^5],
	Sum[G[[k]]^3 * (Sum[G[[l]], {l, k}] - G[[k]]/2)^2, {k, Length[G]}]
} == 0, Reals][[-1]]//RootReduce
G /. N[sol, 40]
```

Steven-Roberts commented on src/arkode/arkode_sprkstep.c:
```diff
@@ -0,0 +1,781 @@
+    sunrealtype bi = step_mem->method->b[is];
+
+    /* Set current stage time(s) */
+    ark_mem->tcur = ark_mem->tn + ai * ark_mem->h;
```
Wouldn't the abscissa be $t_n + \sum_{j} a_j$?

balos1 replied:
Yes, fixed.

Steven-Roberts commented on src/arkode/arkode_sprkstep.c:
```diff
@@ -0,0 +1,781 @@
+    step_mem->istage = is;
+
+    /* set current stage time(s) */
+    ark_mem->tcur = ark_mem->tn + method->b[is] * ark_mem->h;
```
This uses b for the abscissa where before it used a. Does this matter?

balos1 replied:
There was an inconsistency, although it doesnt matter for the problems we currently support.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1417 @@
+
+   :param arkode_mem: pointer to the SPRKStep memory block.
+   :param hfun: name of user-supplied adaptivity function.
+   :param h_data: pointer to user data passed to *hfun* every time it is called.
```
This function should be removed actually, but it is also not the same as the `user_data` passed to the RHS function.

balos1 commented on doc/arkode/guide/source/ARKodeSPRKStorage.rst:
```diff
@@ -0,0 +1,223 @@
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticPseudoLeapfrog2()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic Pseudo Leapfrog 2-2 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticRuth3()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic Ruth 3-3 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticCandyRozmus4()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic Candy Rozmus 4-4 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticMcLachlan2()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic McLachlan 2-2 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticMcLachlan3()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic McLachlan 3-3 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticMcLachlan4()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic McLachlan 4-4 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticMcLachlan5()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic McLachlan 5-6 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticYoshida6()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic Yoshida 6-8 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticMcLachlan8()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic McLachlan 8-16 method.
+
+.. c:function:: ARKodeSPRKStorage ARKodeSymplecticSofroniou10()
+
+   Create the ARKodeSPRKStorage structure for the Symplectic Sofroniou 10-36 method.
```
I think we can remove them from the API, but they are useful internally to avoid duplicate code. The alternative is to use the macro approach we use for the Butcher tables, but I opted not to do that here since there are not nearly as many methods.

balos1 commented on doc/arkode/guide/source/Mathematics.rst:
```diff
@@ -460,6 +460,100 @@ simplified form admits a more efficient and memory-friendly implementation
+
+In the default case, the algorithm for a single time-step is as follows
+
+#. Set :math:`P_0 = p_n, Q_1 = q_n`
+
+#. For :math:`i = 1,\ldots,s` do:
+
+   #. :math:`P_i = P_{i-1} + h_{n+1} a_i f_1(Q_i, t_n + C_i h_{n+1})`
+   #. :math:`Q_{i+1} = Q_i + h_{n+1} b_i f_2(P_i)`
+
+#. Set :math:`p_{n+1} = P_s, q_{n+1} = Q_{s+1}`
```
There was inconsistency in my notation and I have changed it to use "c" and "\hat{c}" as well as "a" and "\hat{a}" instead of "a" and "b". The "\hat{c}" are used, but not the "c" right now, but its possible they will be in future extensions.

balos1 commented on doc/arkode/guide/source/Mathematics.rst:
```diff
@@ -460,6 +460,100 @@ simplified form admits a more efficient and memory-friendly implementation
+Since temporal error based adaptive time-stepping is known to ruin the conservation property :cite:p:`HaWa:06`, 
+SPRKStep employs a fixed time-step size by default. However, it is possible for a user to provide a
+problem-specific adaptivity controller such as the one described in :cite:p:`HaSo:05`.
+The `ark_kepler.c` example demonstrates an implementation of such controller. 
```
Commented out the controller part for now.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/Skeleton.rst:
```diff
@@ -0,0 +1,130 @@
+   this memory structure. See :numref:`ARKODE.Usage.SPRKStep.Initialization` for
+   details.
+
+#. Specify time step size or adaptivity module 
+
+   Call :c:func:`SPRKStepSetFixedStep()` or
+   :c:func:`SPRKStepAdaptivityFn()` to specify either a fixed time-step 
+   size or a callback function that adapts the time-step size. SPRKStep
+   does not support error-based adaptivity like other ARKODE time-stepper 
+   modules due to the incompatibility with symplectic methods.
+
```
Commented out controller parts for now.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1462 @@
+Optional inputs for time step adaptivity
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+SPRKStep only supports a custom time step adaptivity function. The adaptivity methods
+described in :numref:`ARKODE.Mathematics.Adaptivity` are not compatible.
```
Later on when we add adaptivity back in ... yes.

balos1 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,824 @@
+    args.dt                     = SUN_RCONST(1e-3);
+    args.step_mode              = 0;
+    args.stepper                = 1;
+    args.method_name            = "ARKODE_ARK548L2SAb_ERK_8_4_5";
```
The high order ERK with the time-step used keeps the conservation error low enough for the time period simulated.  I get very similar order results using the highest order SPRK to generate the reference instead.

balos1 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,824 @@
+    }
+    else
+    {
+      fprintf(stderr,
+              "ERROR: adaptive time-steps are not supported with SPRKStep\n");
+      return 1;
```
Its true until we add the new adaptivity module stuff in.

balos1 commented on doc/arkode/guide/source/ARKodeSPRKStorage.rst:
```diff
@@ -0,0 +1,223 @@
+
+The following enum values are used to identify different SPRK methods:
+
+.. c:macro:: ARKODE_SPRK_NONE
+
+   Identifier representing no SPRK method, this is solely used to mark the beginning of the enum.
```
I think documenting it is better than not, since its in a public header file.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1417 @@
+      * *ARK_ILL_INPUT* if an argument has an illegal value
+
+   **Notes:**
+      Pass 0.0 to return SPRKStep to the default (adaptive-step) mode.
+
+      Use of this function is not generally recommended, since it gives no
+      assurance of the validity of the computed solutions.  It is
+      primarily provided for code-to-code verification testing purposes.
```
Yep, good catch.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1462 @@
+for the previous problem.  This condition is automatically fulfilled
+if the method order *q* is left unchanged.
+
+One important use of the :c:func:`SPRKStepReInit()` function is in the
+treating of jump discontinuities in the RHS function.  Except in cases
+of fairly small jumps, it is usually more efficient to stop at each
+point of discontinuity and restart the integrator with a readjusted
+ODE model, using a call to this routine.  To stop when the location
+of the discontinuity is known, simply make that location a value of
+``tout``.  To stop when the location of the discontinuity is
+determined by the solution, use the rootfinding feature.  In either
+case, it is critical that the RHS function *not* incorporate the
+discontinuity, but rather have a smooth extension over the
+discontinuity, so that the step across it (and subsequent rootfinding,
+if used) can be done efficiently.  Then use a switch within the RHS
+function (communicated through ``user_data``) that can be flipped
+between the stopping of the integration and the restart, so that the
+restarted problem uses the new values (which have jumped).  Similar
+comments apply if there is to be a jump in the dependent variable
+vector.
```
A brief literature review (mainly a quick read of [this paper](https://www.sciencedirect.com/science/article/pii/S0021999121007415?casa_token=B-3PdJGAnYwAAAAA:irVYQ5VxxDyTLhdEDvTakvQEV7MqlEqw9WppZVLKt-tqeqPKIexe1nijYeAdvzSCv2xEALHrbg)) seems to indicate that jump discontinuities could occur in Hamiltonian problems that are physically interesting. However, its less clear how well (if at all) this approach would work for these problems. SHould we leave the comment in for now and then maybe do an experiment in the future? Or should we leave the comment out (and still do a future experiment)?

drreynolds replied:
Since we are less confident about this use case, then perhaps we can leave it in with some caveats, and include a reference to that citation.  Perhaps something like:

One potential use of the :c:func:`SPRKStepReInit()` function is to treat jump discontinuities in the RHS function [citation].  In lieu of including `if` statements within the RHS function to handle discontinuities, it may be more computationally efficient to stop at each point of discontinuity (e.g., through use of ``tout`` or the rootfinding feature) and restart the integrator with a readjusted ODE model, using a call to this routine.  We note that for the solution to retain temporal accuracy, the RHS function should *not* incorporate the discontinuity.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1462 @@
+SPRKStep system resize function
+-------------------------------------
+
+For simulations involving changes to the number of equations and
+unknowns in the ODE system (e.g. when using spatially-adaptive
+PDE simulations under a method-of-lines approach), the SPRKStep
+integrator may be "resized" between integration steps, through calls
+to the :c:func:`SPRKStepResize()` function. This function modifies
+SPRKStep's internal memory structures to use the new problem size,
+without destruction of the temporal adaptivity heuristics.  It is
+assumed that the dynamical time scales before and after the vector
+resize will be comparable, so that all time-stepping heuristics prior
+to calling :c:func:`SPRKStepResize()` remain valid after the call.  If
+instead the dynamics should be recomputed from scratch, the SPRKStep
+memory structure should be deleted with a call to
+:c:func:`SPRKStepFree()`, and recreated with a call to
+:c:func:`SPRKStepCreate`.
```
I removed the function as it does not seem like we would need it. We can always add it in later if the use case arises.

balos1 commented on src/arkode/arkode.c:
```diff
@@ -1657,6 +1657,10 @@ booleantype arkAllocVectors(ARKodeMem ark_mem, N_Vector tmpl)
   if (!arkAllocVec(ark_mem, tmpl, &ark_mem->yn))
     return(SUNFALSE);
 
+  /* Allocate yerr if needed */
+  if (!arkAllocVec(ark_mem, tmpl, &ark_mem->yerr))
+    return(SUNFALSE);
+
```
I moved it to SPRKStep. It was there as leftover from an early draft implementation.

drreynolds commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1309 @@
+
+  +-----------------------------------------------------+--------------------------------------------+
+  | Optional output                                     | Function name                              |
+  +-----------------------------------------------------+--------------------------------------------+
```
```suggestion
  +=====================================================+============================================+
```

drreynolds commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1309 @@
+
+   +--------------------------------------------------+---------------------------------+
+   | Optional output                                  | Function name                   |
+   +--------------------------------------------------+---------------------------------+
```
```suggestion
   +==================================================+=================================+
```

drreynolds commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1309 @@
+
+   +----------------------------------------+--------------------------------------+
+   | Optional routine                       | Function name                        |
+   +----------------------------------------+--------------------------------------+
```
```suggestion
   +========================================+======================================+
```

drreynolds commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1309 @@
+To set a different step size or have SPRKStep estimate a new step size use
+:c:func:`SPRKStepSetInitStep()`.
+
+One important use of the :c:func:`SPRKStepReset()` function is in the
+treating of jump discontinuities in the RHS functions.  Except in cases
+of fairly small jumps, it is usually more efficient to stop at each
+point of discontinuity and restart the integrator with a readjusted
+ODE model, using a call to :c:func:`SPRKStepReset()`.  To stop when
+the location of the discontinuity is known, simply make that location
+a value of ``tout``.  To stop when the location of the discontinuity
+is determined by the solution, use the rootfinding feature.  In either
+case, it is critical that the RHS functions *not* incorporate the
+discontinuity, but rather have a smooth extension over the
+discontinuity, so that the step across it (and subsequent rootfinding,
+if used) can be done efficiently.  Then use a switch within the RHS
+functions (communicated through ``user_data``) that can be flipped
+between the stopping of the integration and the restart, so that the
+restarted problem uses the new values (which have jumped).  Similar
+comments apply if there is to be a jump in the dependent variable
+vector.
```
I recommend that we just remove this paragraph, since it matches what was said above for ReInit, and doesn't really apply for just resetting an initial condition.

drreynolds commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,820 @@
+
+#include <arkode/arkode_arkstep.h> /* prototypes for ARKStep fcts., consts */
+#include <arkode/arkode_sprk.h>
+#include <arkode/arkode_sprkstep.h> /* prototypes for MRIStep fcts., consts */
```
```suggestion
#include <arkode/arkode_sprkstep.h> /* prototypes for SPRKStep fcts., consts */
```

drreynolds commented on src/arkode/arkode_sprkstep_io.c:
```diff
@@ -0,0 +1,478 @@
+  return (arkSetMaxHnilWarns(arkode_mem, mxhnil));
+}
+
+int SPRKStepSetInitStep(void* arkode_mem, realtype hin)
+{
+  return (arkSetInitStep(arkode_mem, hin));
+}
+
```
As I mentioned in the documentation, I recommend that this be removed from the public UI until time adaptivity is enabled.

drreynolds commented on include/arkode/arkode_sprkstep.h:
```diff
@@ -0,0 +1,127 @@
+                                            void* h_data);
+SUNDIALS_EXPORT int SPRKStepSetMaxNumSteps(void* arkode_mem, long int mxsteps);
+SUNDIALS_EXPORT int SPRKStepSetMaxHnilWarns(void* arkode_mem, int mxhnil);
+SUNDIALS_EXPORT int SPRKStepSetInitStep(void* arkode_mem, realtype hin);
```
As I mentioned in the documentation, I recommend that this routine be removed from the public UI until time adaptivity is enabled.

drreynolds commented on include/arkode/arkode_sprkstep.h:
```diff
@@ -0,0 +1,127 @@
+SUNDIALS_EXPORT int SPRKStepSetInterpolantDegree(void* arkode_mem, int degree);
+/* TODO(CJB): should we remove this from the initial release and wait for the OO
+ * adaptivity? */
+SUNDIALS_EXPORT int SPRKStepSetAdaptivityFn(void* arkode_mem, ARKAdaptFn hfun,
+                                            void* h_data);
```
I recommend that this be removed from the public UI until time adaptivity is enabled.

drreynolds commented on include/arkode/arkode_sprkstep.h:
```diff
@@ -0,0 +1,127 @@
+SUNDIALS_EXPORT int SPRKStepSetAdaptivityFn(void* arkode_mem, ARKAdaptFn hfun,
+                                            void* h_data);
+SUNDIALS_EXPORT int SPRKStepSetMaxNumSteps(void* arkode_mem, long int mxsteps);
+SUNDIALS_EXPORT int SPRKStepSetMaxHnilWarns(void* arkode_mem, int mxhnil);
```
I recommend that this be removed from the public UI until time adaptivity is enabled.

drreynolds commented on src/arkode/arkode_sprkstep_io.c:
```diff
@@ -0,0 +1,478 @@
+  return (arkSetMaxNumSteps(arkode_mem, mxsteps));
+}
+
+int SPRKStepSetMaxHnilWarns(void* arkode_mem, int mxhnil)
+{
+  return (arkSetMaxHnilWarns(arkode_mem, mxhnil));
+}
+
```
I recommend that this be removed from the public UI until time adaptivity is enabled.

drreynolds reviewed (changes requested):
This is looking really great.  I have only a few more items to resolve before I think this is ready.

Steven-Roberts commented on src/arkode/arkode_sprk.c:
```diff
@@ -0,0 +1,574 @@
+/*
+  The following methods are from:
+
+  McLachlan, R.I.: On the Numerical Integration of Ordinary Differential
+  Equations by Symmetric Composition Methods. Siam J Sci Comput. 16, 151–168
+  (1995). https://doi.org/10.1137/0916010
```
According to Geometric Numerical Integration pg. 57, this method is originally due to M. Suzuki & K. Umeno https://doi.org/10.1007/978-3-642-78448-4_7

balos1 replied:
Thanks! I'll add the reference and change the name in the code to SuzukiUmeno (if they found it first, I think the right thing to do is use their name).

drreynolds commented on doc/arkode/guide/source/ARKodeSPRKStorage.rst:
```diff
@@ -0,0 +1,177 @@
+
+   +----------------------------------------------+------------------------------------------------------------+
+   | **Function name**                            | **Description**                                            |
+   +----------------------------------------------+------------------------------------------------------------+
```
```suggestion
   +===============================================+============================================================+
```

drreynolds reviewed (approved):
One minor formatting comment, but otherwise this looks good to me.

balos1 commented:
@drreynolds  I have added a third example to demonstrate handling non-autonomous problems and I have updated the docs accordingly (mainly the math section).  This was spurred by discussion @Steven-Roberts and I had offline. 

I think other then regenerating output files this is good to go now, but Id appreciate if both of you look over the non-autonomous case now.

drreynolds commented on doc/arkode/guide/source/Mathematics.rst:
```diff
@@ -562,6 +569,63 @@ conservation property :cite:p:`HaWa:06`,  SPRKStep employs a fixed time-step siz
+In this non-autonomous case, the Hamiltonian no longer yields a conserved quantity (such as energy)
+although other invariants may exist :cite:p:`Struckmeier:02`.
+However, symplectic methods still are benefical to use for the long-time integration of these systems.
+For example, they may accurately capture the bounded energy of the system :cite:p:``.
```
Missing citation

balos1 replied:
Fixed in the re-written section. I remove this non-autonomous section as I realized that with a minor tweak in TakeStep we do not actually need the extended state-space.  Instead, I generalized the whole section to nonautonomous.

drreynolds commented on examples/arkode/C_serial/ark_anharmonic_symplectic.c:
Is there a `.out` file for this new example (it's listed in `scripts/arkode` but doesn't seem to have been added to the repository)

balos1 replied:
Fixed

drreynolds commented on examples/arkode/C_serial/ark_anharmonic_symplectic.c:
```diff
@@ -0,0 +1,307 @@
+  /* Do integration */
+  for (iout = 0; iout < num_output_times; iout++)
+  {
+    SPRKStepSetStopTime(arkode_mem, tout);
```
It might be nice if the program included a command-line argument to enable/disable "tstop" mode (like you do for compensated sums), so that it would be easy to test the effects of interpolation.  However, this could be a "someday to-do" and shouldn't hold up the PR.

balos1 replied:
Added in each of the examples

drreynolds reviewed (commented):
A couple of minor issues, but otherwise this looks good to go.

balos1 commented:
I think all comments are addressed now. I just need to update the out files for the right machines.

drreynolds reviewed (dismissed):
LGTM

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+
+      Dense outputs may or may not conserve the Hamiltonian. Our testing has
+      shown that Lagrange interpolation typically performs well in this regard,
+      while Hermite interpolation does not. 
```
I think this version addresses by comments above about clarifying what is meant by better.

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+   :numref:`ARKODE.Usage.SPRKStep.SPRKStepRootOutputs`,
+
+#. General usability routines (e.g. to print the current SPRKStep
+   parameters, or output the current Butcher table) are in
```
```suggestion
   parameters, or output the current Butcher tables) are in
```

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+
+* The ratio *nsteps/step_attempts* can measure the quality of the
+  time step adaptivity algorithm, since a poor algorithm will result
+  in more failed steps, and hence a lower ratio.
```
Comment out for now

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+  +-----------------------------------------------------+--------------------------------------------+
+  | No. of calls to right-hand side functions           | :c:func:`SPRKStepGetNumRhsEvals()`         |
+  +-----------------------------------------------------+--------------------------------------------+
+  | Current method memory                               | :c:func:`SPRKStepGetCurrentMethod()`       |
```
```suggestion
  | Current method table                                | :c:func:`SPRKStepGetCurrentMethod()`       |
```

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+      * *ARK_SUCCESS* if successful
+      * *ARK_MEM_NULL* if the ARKStep memory was ``NULL``
+
+   **Notes:**
```
Should probably upgrade this to `.. warning::`

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+
+.. c:function:: char *SPRKStepGetReturnFlagName(long int flag)
+
+   Returns the name of the SPRKStep constant corresponding to *flag*.
```
Adding reference to the table of constants would be helpful here

balos1 replied:
Indeed, Ill ad this in the other steppers too.

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+      * *ARK_MEM_NULL* if the SPRKStep memory was ``NULL``
+
+
+.. c:function:: int SPRKStepGetNumErrTestFails(void* arkode_mem, long int* netfails)
```
Remove?

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+problem must have the same size as the previous one.  This routine
+retains the current settings for all SPRKStep module options and
+performs the same input checking and initializations that are done in
+:c:func:`SPRKStepCreate`, but it performs no memory allocation as is
```
```suggestion
:c:func:`SPRKStepCreate`, but it performs no memory allocation as it
```

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+treating of jump discontinuities in the RHS function :cite:p:`Tao:22`. 
+In lieu of including if statements within the RHS function to handle 
+discontinuities, it may be more computationally efficient to stop at each 
+point of discontinuity (e.g., through use of tout or the rootfinding feature)
```
```suggestion
point of discontinuity (e.g., through use of tstop or the rootfinding feature)
```

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+
+   :param arkode_mem: pointer to the SPRKStep memory block.
+   :param f1: the name of the C function (of type :c:func:`ARKRhsFn()`) defining :math:`f1(q,t) = \frac{\partial V(q,t)}{\partial q}`
+   :param f2:  the name of the C function (of type :c:func:`ARKRhsFn()`) defining :math:`f2(p) = \frac{\partial T(p)}{\partial p}`
```
This notation is inconsistent with the Create function

gardner48 reviewed (changes requested):
Finished a pass over the documentation, I'll get into the examples and source next

gardner48 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,812 @@
+ * The rootfinding feature of SPRKStep is used to count the number of complete orbits. 
+ * This is done by defining the function,
+ *    g(q) = q2
+ * and providing it to SPRKStep as the function to find the roots for.
```
```suggestion
 * and providing it to SPRKStep as the function to find the roots for g(q).
```

gardner48 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,812 @@
+ * The program also accepts command line arguments to change the method
+ * used and time-stepping strategy. The program has the following CLI arguments:
+ * 
+ *   --step-mode <fixed, adapt>  should we use a fixed time-step or adaptive time-step (default fixed)
```
Remove for now?

balos1 replied:
Still relevant for ARKStep which is an example option

gardner48 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,812 @@
+ *   --tf <Real>                 the final time for the simulation (default 100)
+ *   --nout                      number of output times
+ *   --count-orbits              use rootfinding to count the number of completed orbits
+ *   --check-order               compute the order of the method used and check if it is within range of the expected 
```
```suggestion
 *   --check-order               compute the order of the method used and check if it is within the expected range
```

gardner48 commented on examples/arkode/C_serial/ark_kepler.c:
I like the option to check the order but I don't think that's necessary to have in a user example. Maybe make this version a unit test and have a simplified version here without the order checking.

balos1 replied:
Personally, I think it's better to have here since we multiple examples (and the others dont have this complication) so that we do not need to maintain two copies of the example. We do a similar thing for an MRIStep example.

gardner48 replied:
It would be good to split up those MRI examples as well or just move them to the `test` directory since they're targeted more at testing than usage. I think this is the only example that uses rootfinding with SPRK step and so I thought a simplified version would be good to have as well.

drreynolds replied:
I'm fine with keeping it in.  

While I agree with David that it's unlikely a typical SUNDIALS user would want to do this kind of sanity check, if a user *did* want to do so then we have some examples of it.  An added benefit is that Cody's version puts the convergence estimation stuff in a separate function (with a clear name), so uninterested users can easily ignore it.

gardner48 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,812 @@
+        fprintf(stdout, "  g[0] = %3d, y[0] = %3Lg, y[1] = %3Lg, num. orbits is now %.2Lf\n",
+                rootsfound, (long double)ydata[0], (long double)ydata[1],
+                (long double)num_orbits);
+        fprintf(stdout, "t = %.4Lf, H(p,q)-H0 = %.16Lf, L(p,q)-L0 = %.16Lf\n",
```
```suggestion
        fprintf(stdout, "t = %.4Lf, H(p,q)-H0 = %.16Le, L(p,q)-L0 = %.16Le\n",
```

gardner48 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,812 @@
+      else if (retval >= 0)
+      {
+        /* Output current integration status */
+        fprintf(stdout, "t = %.4Lf, H(p,q)-H0 = %.16Lf, L(p,q)-L0 = %.16Lf\n",
```
```suggestion
        fprintf(stdout, "t = %.4Lf, H(p,q)-H0 = %.16Le, L(p,q)-L0 = %.16Le\n",
```

gardner48 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,812 @@
+        fprintf(stdout, "  g[0] = %3d, y[0] = %3Lg, y[1] = %3Lg, num. orbits is now %.2Lf\n",
+                rootsfound, (long double)ydata[0], (long double)ydata[1],
+                (long double)num_orbits);
+        fprintf(stdout, "t = %.4Lf, H(p,q)-H0 = %.16Lf, L(p,q)-L0 = %.16Lf\n",
```
```suggestion
        fprintf(stdout, "t = %.4Lf, H(p,q)-H0 = %.16Le, L(p,q)-L0 = %.16Le\n",
```

gardner48 commented on examples/arkode/C_serial/ark_kepler.c:
```diff
@@ -0,0 +1,812 @@
+      else if (retval >= 0)
+      {
+        /* Output current integration status */
+        fprintf(stdout, "t = %.4Lf, H(p,q)-H0 = %.16Lf, L(p,q)-L0 = %.16Lf\n",
```
```suggestion
        fprintf(stdout, "t = %.4Lf, H(p,q)-H0 = %.16Le, L(p,q)-L0 = %.16Le\n",
```

gardner48 commented on examples/arkode/C_serial/ark_damped_harmonic_symplectic.c:
```diff
@@ -0,0 +1,322 @@
+static sunrealtype Hamiltonian(N_Vector yvec, sunrealtype t);
+static int ParseArgs(int argc, char* argv[], ProgramArgs* args);
+static void PrintHelp();
+static int check_retval(void* returnvalue, const char* funcname, int opt);
```
Here and in the other examples I think it would be good to move utilities like `ParseArgs`, `PrintHelp`, `check_retval`, etc. to a header file for the example to streamline what's in the source to just what a user needs to provide.

balos1 replied:
Are you suggesting an example specific header file like `ark_damped_harmonic_symplectic.h`?

gardner48 replied:
Yes, that's what I was thinking (in particular for things like parsing input args that can be lengthy but not that important to understanding how to use the code)

gardner48 commented on examples/arkode/C_serial/ark_kepler_--stepper_SPRK_--step-mode_fixed_--method_ARKODE_SPRK_EULER_1_1_--tf_50_--check-order_--nout_1.out:
Memory leak in example when checking order

gardner48 commented on examples/arkode/C_serial/ark_kepler_--stepper_SPRK_--step-mode_fixed_--method_ARKODE_SPRK_MCLACHLAN_3_3_--tf_50_--check-order_--nout_1.out:
For some reason this test (and others that follow) didn't make it as far as the other before the address sanitizer errored out

balos1 replied:
I like printing more digits of the energy instead because the error is not as clearly defined.

gardner48 commented on examples/arkode/C_serial/ark_kepler_plot.py:
```diff
@@ -0,0 +1,88 @@
+# Programmer(s): Cody J. Balos @ LLNL
+# ----------------------------------------------------------------
+# SUNDIALS Copyright Start
+# Copyright (c) 2002-4022, Lawrence Livermore National Security
```
```suggestion
# Copyright (c) 2002-2023, Lawrence Livermore National Security
```

gardner48 commented on examples/arkode/C_serial/ark_kepler_plot.py:
```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python
```
I think we should use `python3` for new scripts

balos1 replied:
I would agree except that I believe `python` and `python3` are nonstandard and `python` may refer to python version 3 in most modern installations, and some installations (Windows in particular) only have `python` (referring to python version 3).

gardner48 reviewed (changes requested):
Finished a pass over the examples

balos1 commented on doc/arkode/guide/source/Mathematics.rst:
```diff
@@ -460,6 +460,117 @@ simplified form admits a more efficient and memory-friendly implementation
+Hamiltonian for exponentially long times :cite:p:`HaWa:06`. SPRKStep makes the
+assumption that the Hamiltonian is separable, in which case the schemes are
+explicit. SPRKStep provides methods with order of accuracy and conservation
+equal to :math:`q = \{1,2,3,4,5,6,8,10\}`. The tables for these methods, and the
```
The order of conservation of the Hamiltonian is in general the same as the order of accuracy, so this is correct as written.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+
+.. _ARKODE.Usage.SPRKStep.RootFinding:
+
+Rootfinding initialization function
```
I think this is too strong of a statement to make. The Lagrange interpolation seems to do a good job in most cases, and we already mention this in the interpolation section.

gardner48 replied:
Maybe just reiterate that states at root returns are interpolated to the root return time

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+        (b) A root of one of the root functions was found both at a
+            point :math:`t` and also very near :math:`t`.
+
+        (c) The initial condition violates the inequality constraints.
```
No

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+      *tstop* is disabled (and can be re-enabled only though a new call
+      to :c:func:`SPRKStepSetStopTime()`). SPRKStep uses the ARKODE
+      Lagrange interpolation module by default because testing showed that
+      it does a better job of maintaining conservation than Hermite interpolation.
```
> but still introduces additional error compared to the full accuracy solution
OR introduces no additional error.

gardner48 replied:
Somewhere later on in the documentation there was a phrasing I thought was clearer. I don't remember where but I did add a comment that I thought it could address this one (and other similar comments).

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+      * *ARK_MEM_NULL* if the SPRKStep memory is ``NULL``
+      * *ARK_ILL_INPUT* if an argument has an illegal value
+
+   **Notes:**
```
In order to not over use `..note::` and make the documentation cluttered, I dont think we should switch these sections to `..note::` and should just move the stuff into the initial function description.

gardner48 replied:
Agreed, we don't need the `.. note::` section in most cases (only for highlighting particularly important)

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+      accuracy of the time-step solutions on which it is based, the *actual*
+      polynomial degree that is used by SPRKStep will be the minimum of
+      :math:`q-1` and the input *degree*, for :math:`q > 1` where :math:`q` is
+      the order of accuracy for the time integration method.
```
Is this addition from a recent change?

gardner48 replied:
Yes, this was something I added recently in ARKStep and ERKStep where we were getting unexpected results comparing to a native IMEX Euler implementation because of interpolation.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+   +=============================+===========================================+==========+
+   | Set integrator method order | :c:func:`SPRKStepSetOrder()`              | 4        |
+   +-----------------------------+-------------------------------------------+----------+
+   | Set SPRK method             | :c:func:`SPRKStepSetMethod()`             | internal |
+   +-----------------------------+-------------------------------------------+----------+
+   | Set SPRK method by name     | :c:func:`SPRKStepSetMethodName()`         | internal |
```
This follows the ARKStep and ERKStep docs, so for consistency I will update those too.

balos1 commented on doc/arkode/guide/source/Introduction.rst:
```diff
@@ -78,6 +79,17 @@ The algorithms used in ERKStep are adaptive- and fixed-step explicit
+*SPRKStep* focuses on Hamiltonian systems posed in the form,
+
+.. math::
+   H(p, q, t) = T(p) + V(q, t)
```
True, however `p` and `q` are the standard variables to use for these systems so I double people will be confused.

balos1 commented on doc/arkode/guide/source/ARKodeSPRKStorage.rst:
```diff
@@ -0,0 +1,177 @@
+
+   Identifier for the Symplectic Sofroniou 10th order method with 36 stages.
+
+.. c:type:: ARKodeSPRKStorage_s
```
We could have `ARKodeSPRKStorageMem` as the private structure and `ARKodeSPRKStorage` as the public structure to mirror the Butcher table API. The reason I did not use `Table` is to differentiate it more from the Butcher table structure and with the forward thinking to other low storage methods, where it may be awkward to call the data structure a "Table"`.

gardner48 replied:
"Storage" seems a little generic to me. "Table" seemed more natural to me because the coefficients are from a Butcher table but stored an a specialized way to the SPRK case instead of as a generic Butcher table (the same would be true for a particular low storage pattern). We could (and maybe should) similarly have a `ARKodeIMEXTable` type to make it easier to load the tables for an IMEX method.

gardner48 commented on src/arkode/arkode_sprkstep_io.c:
```diff
@@ -0,0 +1,463 @@
+                                  &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  if (ord == 7 || ord == 9 || ord > 10) { return ARK_ILL_INPUT; }
```
This is slightly inconsistent with the documentation. If the user previously changed the order to a valid option then later tries to change the order to an illegal the previously set value will be used instead of the default.

balos1 replied:
So should it behave like the documentation says or should it stay as-is and the documentation updated?

balos1 commented on src/arkode/arkode_sprkstep.c:
```diff
@@ -0,0 +1,786 @@
+  for (is = 0; is < step_mem->method->stages; is++)
+  {
+    /* load/compute coefficients */
+    sunrealtype ai    = step_mem->method->a[is];
+    sunrealtype ahati = step_mem->method->ahat[is];
```
This is not a mixed declaration since its at the top of the block (it passed when I compiled with C90, all warnings, and warnings as errors).

gardner48 replied:
Yes, you're right it's at the top of the block so I think it's ok. It just stood out to me because I think this is the only place where everything is not declared at the top of the function.

drreynolds reviewed (approved):
None of my comments in this re-review are hugely significant, although I am concerned with the `.out` files that show memory leak and/or sanitizer issues.  I'll "approve" this now, in case it's not fully approved before I leave for vacation.

drreynolds commented on src/arkode/arkode_sprkstep.c:
```diff
@@ -0,0 +1,786 @@
+  else { step_mem->yerr = NULL; }
+
+  /* Attach step_mem structure and function pointers to ark_mem */
+  ark_mem->step_attachlinsol   = NULL;
+  ark_mem->step_attachmasssol  = NULL;
+  ark_mem->step_disablelsetup  = NULL;
+  ark_mem->step_disablemsetup  = NULL;
+  ark_mem->step_getlinmem      = NULL;
+  ark_mem->step_getmassmem     = NULL;
+  ark_mem->step_getimplicitrhs = NULL;
+  ark_mem->step_mmult          = NULL;
+  ark_mem->step_getgammas      = NULL;
```
Indeed, `arkCreate` already initializes all of these to `NULL`.

balos1 commented on doc/arkode/guide/source/ARKodeSPRKStorage.rst:
```diff
@@ -0,0 +1,177 @@
+
+.. _SPRKStorage.id:
+
+Method identifiers
```
Should we remove them from Butcher.rst instead of removing them from here since they are separate from the Butcher table structure?

gardner48 replied:
I think leaving these in Butcher.rst make sense so all the method tables are listed in one place (even if they correspond to different data structures for storing the table). They're already under a "Symplectic" section and some introductory text could be add at the start of the section to note the different data structure used in this case. It might also be better to name this section "SPRK Tables" since to, as Dan mentioned somewhere else, distinguish from other symplectic methods with a different table structure.

balos1 commented on examples/arkode/C_serial/ark_harmonic_symplectic.c:
```diff
@@ -0,0 +1,366 @@
+    N_VLinearSum(SUN_RCONST(1.0), y, -SUN_RCONST(1.0), solution, solution);
+    err = sqrt(N_VDotProd(solution, solution));
+
+    /* Output current integration status */
```
Since we output the Energy and it should remain constant, this seems redundant to print out.

gardner48 replied:
There will be some error though since it's not conserved exactly and it's easier to see this in the error (or if we printed more digits of the energy)

balos1 replied:
I like printing more digits of the energy instead because the error is not as clearly defined.

gardner48 replied:
The error would be the difference from the initial energy

balos1 replied:
Well, yes, of course.  But my point is that you can observe conservation and the actual value for the energy just by looking at energy (i.e. error is not necessary, just more digits).

gardner48 replied:
Yes, they're both ways of showing the same thing. I think it's easier to see an error that's something like `1e-15`  but either way is good.

balos1 commented on scripts/arkode:
That would be very nice

balos1 commented on src/arkode/arkode_io.c:
```diff
@@ -159,8 +160,10 @@ int arkSetInterpolantType(void *arkode_mem, int itype)
+  } else if (itype == ARK_INTERP_LAGRANGE) {
     ark_mem->interp = arkInterpCreate_Lagrange(arkode_mem, ARK_INTERP_MAX_DEGREE);
+  } else {
+    ark_mem->interp = NULL;
```
Not now, but I experimented with having a "none" type and I dont think it hurts to have this else branch to catch an error.

balos1 commented on src/arkode/arkode_io.c:
```diff
@@ -1279,6 +1282,31 @@ int arkSetMaxConvFails(void *arkode_mem, int maxncf)
+  }
+  ark_mem = (ARKodeMem) arkode_mem;
+
+  /* argument <= 0 sets default, otherwise set input */
```
Maybe its just not rendering, but this comment appears blank.

gardner48 replied:
The comment is incorrect (any non-zero will enable compensated sums) and the suggested change was to remove the comment

balos1 commented on src/arkode/arkode_sprk.c:
```diff
@@ -0,0 +1,580 @@
+    for (j = 0; j <= i; ++j) { b->A[i][j] = sprk_storage->ahat[j]; }
+  }
+
+  /* Time weights: C_j = sum_{i=0}^{j-1} b_i */
```
I would think that it is not a problem since we are not summing up that many values.

balos1 commented on doc/arkode/guide/source/Butcher.rst:
```diff
@@ -18,12 +18,12 @@
-component Butcher tables are listed within their separate
-sections, but are referenced together in the additive section.
+Here we catalog the full set of Butcher tables included in ARKODE. We group
+these into four categories: *explicit*, *implicit*, *additive* and *symplectic*.
```
Updated the language to be more specific to SPRK.

balos1 commented on doc/arkode/guide/source/ARKodeSPRKStorage.rst:
```diff
@@ -0,0 +1,177 @@
+
+   :param B: The ARKodeSPRKStorage structure to free.
+
+.. c:function:: int ARKodeSPRKStorage_ToButcher(ARKodeSPRKStorage sprk_storage, ARKodeSPRKStorage* a_ptr, ARKodeSPRKStorage* b_ptr)
```
I would prefer not to use this nomenclature in the function signatures to allow room for changes later where the tables are not necessarily ERK or DIRK.

gardner48 replied:
I'm fine with using different names. Note this reflects what's in `sprkstep.h`.

balos1 replied:
Good catch. Thanks, ill update sprkstep.h.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+
+      Since interpolation may reduce the accuracy in the reported
+      solution, if full method accuracy is desired the user should issue
+      a call to :c:func:`SPRKStepSetStopTime()` before the call to
```
I tried to do this, but it resulted in a number of edge cases and then in the future when we add adaptivity we may want to undo it.

gardner48 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+additional values of solution-related quantities.  This function
+should only be called after a successful return from
+:c:func:`SPRKStepEvolve()`, as it provides interpolated values either of
+:math:`y` or of its derivatives (up to the 5th derivative)
```
I don't recall how many of these there were but if it's only a few (hopefully) it might be easier to make these changes at the same time. If there are a bunch of them then we could open an issue with what needs to be fixed and make those changes in a followup PR.

balos1 commented on doc/arkode/guide/source/Usage/SPRKStep_c_interface/User_callable.rst:
```diff
@@ -0,0 +1,1229 @@
+   :param g: name of user-supplied function, of type :c:func:`ARKRootFn()`,
+      defining the functions :math:`g_i` whose roots are sought.
+
+   :return:
```
I went with `:retval VALUE:`

gardner48 reviewed (changes requested):
This looks good, I think all that's needed is to update the answer files for the GitHub CI and once those are merged update the submodule commit here.
